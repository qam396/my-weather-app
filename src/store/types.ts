export interface Weather {
    description: string
    icon: string
    id: string
    main: string
}

export interface Daily {
    weather: Weather[]
    temp: {
        day: number
        min: number
        max: number
        night: number
        eve: number
        morn: number
    }
    dt: number
}

export interface WeatherData {
    base: string
    clouds: {
        all: number
    }
    cod: number
    coord: {
        lon: number
        lat: number
    }
    dt: number
    id: number
    main: {
        feels_like: number
        humidity: number
        pressure: number
        temp: number
        temp_max: number
        tem_main: number
    }
    name: string
    sys: {
        country: string
        id: number
        sunrise: number
        sunset: number
        type: number
    }
    timezone: number
    visibility: number
    weather: Weather[]
    wind: {
        speed: number
        deg: number
    }
    daily: Daily[]
}
