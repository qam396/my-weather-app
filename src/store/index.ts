import { createStore } from 'vuex'
import axios from 'axios'
import { WeatherData } from './types'
import { API_KEY, API_BASE_URL } from '@/helpers/constants'
import createPersistedState from 'vuex-persistedstate'

export default createStore({
  // plugins: [createPersistedState()],
  state: {
    weather: null as WeatherData | null
  },
  mutations: {
    setWeather (state, weather) {
      state.weather = weather
    },
    clearWeather (state) {
      state.weather = null
    }
  },
  actions: {
    getWeather ({ commit }, query) {
      return axios.get(`${API_BASE_URL}weather?q=${query}&units=metric&APPID=${API_KEY}`)
        .then((result) => {
          commit('setWeather', result.data)
        })
        .catch(() => {
          commit('clearWeather')
          throw new Error('Oops.. city not found.')
        })
    },
    dailyForecast () {
      return axios.get(`${API_BASE_URL}onecall?lat=${this.state.weather?.coord.lat}&lon=${this.state.weather?.coord?.lon}&exclude=current,hourly,minutely,alerts&units=metric&appid=${API_KEY}`)
        .then((result) => { return result.data })
        .catch(console.error)
    }
  },
  getters: {
    selectedWeather (state) {
      return state.weather
    }
  }
})
