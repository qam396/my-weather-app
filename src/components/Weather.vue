<template>
  <div>
    <main>
      <!-- component that show loading text -->
      <loader v-if="loading"></loader>
      <h1 class="center header">Weather App</h1>
      <div class="search-box">
        <input
          type="text"
          class="search-bar"
          placeholder="Search by city name..."
          v-model="query"
          @keyup.enter="fetchWeather()" />
      </div>

      <div v-if="error" class="bar error"> {{ error }} </div>

      <div class="weather-wrap" v-if="weather">
        <!-- component that shows location of the search city -->
        <location
          :city="weather.name"
          :country="weather.sys.country"
          :date="currentDate" />
        <!-- component that shows weather related information -->
        <information
          :temperature="Math.round(weather.main.temp)"
          :main="weather.weather[0].main"
          :evening="evening" />
        <div class="center">
          <!-- Navigation link to week details page -->
          <router-link :to="{ name: 'WeatherWeek',
            params: {
                lat: weather.coord.lat,
                lon: weather.coord.lon
            }}" tag="button">View weather for full week</router-link>
        </div>
      </div>
    </main>
  </div>
</template>

<script lang="ts">
import { Options, Vue } from 'vue-class-component'
import { WeatherData } from '../store/types'
import Loader from './shared/Loader.vue'
import Location from './shared/Location.vue'
import Information from './shared/Information.vue'
import { mapGetters, mapActions, mapMutations } from 'vuex'
import { AxiosResponse } from 'axios'

@Options({
  components: {
    Loader,
    Information,
    Location
  },
  computed: {
    // Get selected weather and set it to local variable
    ...mapGetters({ weather: 'selectedWeather' })
  },
  methods: {
    // Map actions and Mutation to use in the component
    ...mapActions(['getWeather']),
    ...mapMutations(['clearWeather'])
  }
})
export default class App extends Vue {
  loading = false
  query = ''
  weather: WeatherData | undefined
  evening = false
  getWeather!: (query: string) => Promise<AxiosResponse<unknown, unknown>>
  clearWeather!: () => Promise<void>
  error = ''

  get currentDate (): string {
    return new Date().toDateString()
  }

  get backgroundImage (): string {
    return this.evening ? 'night-background' : 'day-background'
  }

  /**
   * Call action store to fetch weather and set it to weather state
   * @param body (optional) A new LedgerAggregatedItem to be created
   * @return Created LedgerAggregatedItem
   */
  fetchWeather () : void {
    this.loading = true
    this.error = ''
    this.getWeather(this.query)
      .then(() => this.setResults())
      .catch((error) => {
        this.error = error
        console.log(error)
      })
      .finally(() => { this.loading = false })
  }

  /**
   * Use weather icon and set evening if icon contains n 'night' 
   * @param body (optional) A new LedgerAggregatedItem to be created
   * @return Created LedgerAggregatedItem
   */
  setResults () : void {
    const timeOfDay = this.weather?.weather[0].icon
    if (timeOfDay?.includes('n')) {
      this.evening = true
    }
  }
}
</script>

<style>
  .search-box {
    width: 100%;
    margin-bottom: 30px;
  }
  .search-box .search-bar {
    display: block;
    width: 100%;
    padding: 15px;
    color: #313131;
    font-size: 20px;

    appearance: none;
    border: none;
    outline: none;
    background: none;

    box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.25);
    background-color: rgba(255, 255, 255, 0.5);
    border-radius: 0px 16px 0px 16px;
    transition: 0.4s;
  }
  .search-box .search-bar:focus {
    box-shadow: 0px 0px 16px rgba(0, 0, 0, 0.25);
    background-color: rgba(255, 255, 255, 0.75);
    border-radius: 16px 0px 16px 0px;
  }
  .center {
    text-align: center;
  }
  .day-background {
    background-image: url('../assets/background-day.jpg');
    background-size: cover;
    background-position: center;
    margin-left: 500px;
    margin-right: 500px;
    border-radius: 16px;
  }
  .night-background {
    background-image: url('../assets/background-night.jpg');
    background-size: cover;
    background-position: center;
    margin-left: 500px;
    margin-right: 500px;
    border-radius: 16px;
  }
  .bar {
    padding: 10px;
    margin: 10px;
    color: #333;
    background: #fafafa;
    border: 1px solid #ccc;
  }
  .error {
    color: #ba3939;
    background: #ffe0e0;
    border: 1px solid #a33a3a;
  }
  .header {
    color: white;
    padding-bottom: 30px;
  }
</style>
