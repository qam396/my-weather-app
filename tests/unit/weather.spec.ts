import { WeatherData } from '@/store/types'
import { shallowMount, VueWrapper } from '@vue/test-utils'
import Weather from 'src/components/Weather.vue'
import { Vue } from 'vue-class-component'
import Vuex from 'vuex'

describe('weather page', () => {
  let wrapper: VueWrapper<Vue>
  let vm: any

  beforeEach(async () => {
    const state = {
      weather: () => { return { name: 'london' } as WeatherData }
    }

    const getters = {
      selectedWeather: () => { return state.weather }
    }

    const storeMockData = new Vuex.Store({
      modules: {
        projects: {
          namespaced: true,
          state,
          getters
        }
      }
    })

    wrapper = shallowMount(Weather, {
      props: {},
      global: {
        plugins: [storeMockData]
      }
    })

    if (wrapper !== undefined) {
      vm = wrapper.vm
      vm.loading = false
    }
  })

  it('should passes the sanity check and create wrapper', () => {
    expect(wrapper).toBeTruthy()
  })

  it('should return weather back', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have a one search box', () => {
    expect(wrapper.findAll('.search-box')).toHaveLength(1)
  })

  it('should not weather wrap unless weather is defined', () => {
    expect(wrapper.findAll('.weather-wrap')).toHaveLength(0)
  })

  it('should show loading text', () => {
    vm.loading = false
    vm.test()
    expect(vm.loading).toBeTruthy()
  })
})
