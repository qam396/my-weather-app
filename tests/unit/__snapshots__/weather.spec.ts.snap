// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weather page should return weather back 1`] = `
<div>
  <main>
    <!--v-if-->
    <div
      class="search-box"
    >
      <input
        class="search-bar"
        placeholder="Search..."
        type="text"
      />
    </div>
    <!--v-if-->
    <!--v-if-->
  </main>
</div>
`;
